### PR TITLE
(CDAP-14984) Introduce master environment task

### PR DESCRIPTION
**Pull request by chtyim**
- Also implemented a pod killer to periodically kill the preview pod

### CODE REVIEW

File: cdap-master/src/main/java/co/cask/cdap/master/environment/k8s/MasterTaskExecutorService.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A service for executing {@link MasterEnvironmentTask} periodically.
+ */
+final class MasterTaskExecutorService extends AbstractScheduledService {
```
Comment (CuriousVini) [inline cdap-master/src/main/java/co/cask/cdap/master/environment/k8s/MasterTaskExecutorService.java:33]: should we extend from `AbstractRetryableScheduledService` instead?

File: cdap-kubernetes/src/main/java/co/cask/cdap/master/environment/k8s/PodKillerTask.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    try {
+      CoreV1Api api = getCoreApi();
+      LOG.debug("Terminating pods using selector {}", podSelector);
+      api.deleteCollectionNamespacedPod(namespace, null, null, null, null,
+                                        podSelector, null, null, null, null);
+      LOG.debug("Pods termination completed");
+    } catch (IOException e) {
+      LOG.warn("IO Exception raised when connecting to Kubernetes API server", e);
+    } catch (ApiException e) {
+      LOG.warn("API exception raised when trying to delete pods, code=" + e.getCode()
```
Comment (CuriousVini) [inline cdap-kubernetes/src/main/java/co/cask/cdap/master/environment/k8s/PodKillerTask.java:58]: nit: `Api Exception` to be consistent with `IO Exception`. Or we can just use `Exception raised. ..`

File: cdap-kubernetes/src/main/java/co/cask/cdap/master/environment/k8s/PodKillerTask.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+                 + ", body=" + e.getResponseBody(), e);
+    }
+
+    return delayMillis;
```
Comment (CuriousVini) [inline cdap-kubernetes/src/main/java/co/cask/cdap/master/environment/k8s/PodKillerTask.java:62]: we should update `delayMillis` to a negative number to terminate the task right?

**Review by CuriousVini — COMMENTED** [review]
A few comments.

File: cdap-master/src/main/java/co/cask/cdap/master/environment/k8s/MasterTaskExecutorService.java (hunk 1 above)
Comment (chtyim) [reply cdap-master/src/main/java/co/cask/cdap/master/environment/k8s/MasterTaskExecutorService.java:33]: Why? We only need simply infinite retry in here.

File: cdap-kubernetes/src/main/java/co/cask/cdap/master/environment/k8s/PodKillerTask.java (hunk 2 above)
Comment (chtyim) [reply cdap-kubernetes/src/main/java/co/cask/cdap/master/environment/k8s/PodKillerTask.java:58]: They are different exception. The IOException is caused by connection issue. The ApiException means the call returns some error code.

File: cdap-kubernetes/src/main/java/co/cask/cdap/master/environment/k8s/PodKillerTask.java (hunk 3 above)
Comment (chtyim) [reply cdap-kubernetes/src/main/java/co/cask/cdap/master/environment/k8s/PodKillerTask.java:62]: No. We want the task keeps getting invoked.

File: cdap-master/src/main/java/co/cask/cdap/master/environment/k8s/MasterTaskExecutorService.java (hunk 1 above)
Comment (CuriousVini) [reply cdap-master/src/main/java/co/cask/cdap/master/environment/k8s/MasterTaskExecutorService.java:33]: I see.

**Review by CuriousVini — APPROVED** [review]
lgtm 👍

**Comment by chtyim** [comment]
Build https://builds.cask.co/browse/CDAP-DUT6887

**Pull request merged**